File: services/ms-notes/src/main/java/org/medilabo/msnotes/service/NoteService.java
package org.medilabo.msnotes.service;

import org.medilabo.msnotes.document.Note;
import org.medilabo.msnotes.dto.NoteDto;
import org.medilabo.msnotes.model.GenericResponseModel;

import java.util.List;

public interface NoteService {

    /**
     * Retrieves all notes in the system.
     *
     * @return a list of all notes
     */
    List<Note> findAllNotes();

    /**
     * Finds a note by its unique identifier.
     *
     * @param id the note identifier
     * @return the note if found
     */
    Note findById(String id);

    /**
     * Retrieves all notes associated with a specific patient.
     *
     * @param patientId the patient identifier
     * @return a list of notes for the specified patient
     */
    List<Note> findAllByPatientId(int patientId);

    /**
     * Creates a new note in the system.
     *
     * @param noteDto the note data transfer object
     * @return the created note
     */
    Note addNote(NoteDto noteDto);

    /**
     * Deletes a note from the system.
     *
     * @param id the note identifier to delete
     * @return a response indicating success or failure of the operation
     */
    GenericResponseModel deleteNote(String id);
}

File: services/ms-ui/src/app/services/note.service.ts
import {inject, Injectable} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {Observable} from 'rxjs';
import {Note} from '../models/note.model';
import {environment} from '../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class NoteService {
  private http = inject(HttpClient);

  private readonly apiUrl = environment.api.gateway_url + '/note';

  /**
   * Retrieves all medical notes for a specific patient
   *
   * @param patientId - The unique identifier of the patient
   * @returns An Observable containing an array of Note objects
   */
  getNotesByPatientId(patientId: Number): Observable<Note[]> {
    return this.http.get<Note[]>(`${this.apiUrl}/patient/${patientId}`);
  }

  /**
   * Creates a new medical note
   *
   * @param note - The Note object to be created
   * @returns An Observable containing the created Note object with assigned ID
   */
  addNote(note: Note): Observable<Note> {
    return this.http.post<Note>(`${this.apiUrl}`, note);
  }

  /**
   * Deletes a specific medical note
   *
   * @param noteId - The unique identifier of the note to delete
   * @returns An Observable that completes when the deletion is successful
   */
  deleteNote(noteId: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${noteId}`);
  }
}
